src/shims/fs.rs

```diff
@@ -257,6 +264,197 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+        let flags = this.read_scalar(flags_op)?.to_machine_isize(&*this.tcx)? as i32;
+
+        let at_empty_path = this.eval_libc_i32("AT_EMPTY_PATH")?;
+        let at_fdcwd = this.eval_libc_i32("AT_FDCWD")?;
+
+        // Now we need to resolve the path of the target file. This was done following the manual
+        // entry for `statx`
+
+        // If `pathname` is absolute we can ignore `dirfd`.
+        let path = if pathname.is_absolute() {
+            pathname
+        } else {
+            let dirfd = this.read_scalar(dirfd_op)?.to_machine_isize(&*this.tcx)? as i32;
+            // If `dirfd` is `AT_FDCWD`, `pathname` is relative to the current directory.
+            if dirfd == at_fdcwd {
+                match std::env::current_dir() {
+                    Ok(cwd) => cwd.join(pathname),
+                    Err(e) => {
+                        this.set_last_error_from_io_error(e)?;
+                        return Ok(-1);
+                    }
+                }
+            } else {
+                if let Some(path) = &this
+                    .machine
+                    .file_handler
+                    .handles
+                    .get(&dirfd)
+                    .map(|h| &h.path)
+                {
+                    // If `dirfd` points to a directory, `pathname` is relative to that directory
+                    if path.is_dir() {
+                        path.join(pathname)
+                    } else {
+                        // If `pathname` is empty and the `AT_EMPTY_PATH` flag is set, we use the path of `dirfd`.
+                        if pathname.as_os_str().is_empty() && (flags & at_empty_path != 0) {
+                            (*path).clone()
+                        } else {
+                            // It is not specified what should happen here.
+                            throw_unsup_format!(
+                                "the file descriptor `{}` and path `{:?}` combination is invalid.",
+                                dirfd,
+                                path
+                            );
+                        }
+                    }
+                } else {
+                    return this.handle_not_found();
+                }
+            }
+        };
+
+        // the `_mask_op` paramter specifies the file information that the caller requested.
+        // However `statx` is allowed to return information that was not requested or to not
+        // return information that was requested. This `mask` represents the information we can
+        // actually provide in any host platform.
+        let mut mask =
+            this.eval_libc("STATX_TYPE")?.to_u32()? | this.eval_libc("STATX_SIZE")?.to_u32()?;
+
+        // If the `AT_SYMLINK_NOFOLLOW` flag is set, we query the file's metadata without following
+        // symbolic links.
+        let metadata = if flags & this.eval_libc("AT_SYMLINK_NOFOLLOW")?.to_i32()? != 0 {
```

Comment:
Here's another branch where both sides need testing.

Reply:
I just checked and we need `libc::symlink` for creating symbolic links from miri. I'll add a `FIXME` and implement it ASAP.